src/sagetasks/nextflowtower/utils.py

```diff
@@ -0,0 +1,177 @@
+        assert platform is None or platform in ENDPOINTS
+        assert platform or endpoint
+        endpoint = endpoint if endpoint else ENDPOINTS[platform]
+        client_args = dict(tower_token=auth_token, tower_api_url=endpoint, **kwargs)
+        return client_args
+
+    def get_compute_env(self, compute_env_id):
+        endpoint = f"/compute-envs/{compute_env_id}"
+        params = self.init_params()
+        response = self.client.request("GET", endpoint, params=params)
+        compute_env = self.extract_resource(response)
+        return compute_env
+
+    def get_workflow(self, workflow_id):
+        endpoint = f"/workflow/{workflow_id}"
+        params = self.init_params()
+        response = self.client.request("GET", endpoint, params=params)
+        return response
+
+    @staticmethod
+    def update_dict(base_dict, overrides):
+        dict_copy = copy(base_dict)
+        for k, v in overrides.items():
+            oldv = dict_copy.get(k, {})
+            if k not in dict_copy:
+                valid = set(dict_copy)
+                raise ValueError(f"Cannot update {k}. Not among {valid}.")
+            elif isinstance(oldv, Mapping) and isinstance(v, Mapping):
+                dict_copy[k] = TowerUtils.update_dict(oldv, v)
+            elif v is not None:
+                dict_copy[k] = v
+        return dict_copy
+
+    def init_launch_workflow_data(self, compute_env_id):
+        # Retrieve compute environment for default values
+        compute_env = self.get_compute_env(compute_env_id)
+        assert compute_env["status"] == "AVAILABLE"
+        ce_config = compute_env["config"]
+        # Replicating date format in requests made by Tower frontend
+        now_utc = datetime.now().isoformat()[:-3] + "Z"
+        data = {
+            "launch": {
+                "computeEnvId": compute_env_id,
+                "configProfiles": [],
+                "configText": None,
+                "dateCreated": now_utc,
+                "entryName": None,
+                "id": None,
+                "mainScript": None,
+                "paramsText": None,
+                "pipeline": None,
+                "postRunScript": ce_config["postRunScript"],
+                "preRunScript": ce_config["preRunScript"],
+                "pullLatest": None,
+                "revision": None,
+                "runName": None,
+                "schemaName": None,
+                "stubRun": None,
+                "towerConfig": None,
+                "userSecrets": [],
+                "workDir": ce_config["workDir"],
+                "workspaceSecrets": [],
+            }
+        }
+        return data
+
+    def launch_workflow(
+        self,
+        compute_env_id,
+        pipeline,
+        revision=None,
+        params_yaml=None,
+        params_json=None,
+        nextflow_config=None,
+        run_name=None,
+        work_dir=None,
+        profiles=(),
+        workspace_secrets=(),
+        user_secrets=(),
```

Comment:
I don't know. It's hard to test all possible permutations with complex requests like this one. I added a `dedup()` utility function to avoid the situation.